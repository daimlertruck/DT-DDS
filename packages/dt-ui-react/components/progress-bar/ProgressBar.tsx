import { Typography } from '@dt-ui/react-typography';

import { CheckCircleOutlineIcon, ErrorOutlineIcon } from '../../core';

import { MAX_PROGRESS_BAR_VALUE, MIN_PROGRESS_BAR_VALUE } from './constants';
import {
  ProgressBarFieldStyled,
  ProgressBarWrapperStyled,
  ProgressBarStyled,
  ProgressBarTrackStyled,
  ProgressBarIconStyled,
  ProgressBarPercentageTextStyled,
  ProgressBarHelperTextStyled,
} from './ProgressBar.styled';
import { ProgressBarSize, ProgressBarState, ProgressBarInfo } from './types';

export interface ProgressBarProps {
  state: ProgressBarState;
  value: number;
  label?: string;
  helperText?: string;
  info?: ProgressBarInfo;
  size?: ProgressBarSize;
}

export const ProgressBar = ({
  state,
  value,
  label,
  helperText,
  size = 'small',
  info = 'none',
}: ProgressBarProps) => {
  const normalizedValue = Math.min(
    Math.max(value, MIN_PROGRESS_BAR_VALUE),
    MAX_PROGRESS_BAR_VALUE
  );
  return (
    <ProgressBarFieldStyled>
      {label ? (
        <Typography color='content.body' fontStyles='body2Bold'>
          {label}
        </Typography>
      ) : null}

      <ProgressBarWrapperStyled>
        <ProgressBarTrackStyled
          aria-label={label}
          aria-valuemax={MAX_PROGRESS_BAR_VALUE}
          aria-valuemin={MIN_PROGRESS_BAR_VALUE}
          aria-valuenow={normalizedValue}
          role='progressbar'
        >
          <ProgressBarStyled
            progress={normalizedValue}
            size={size}
            state={state}
          />
        </ProgressBarTrackStyled>

        {renderInfo()}
      </ProgressBarWrapperStyled>

      {helperText ? (
        <ProgressBarHelperTextStyled state={state}>
          {helperText}
        </ProgressBarHelperTextStyled>
      ) : null}
    </ProgressBarFieldStyled>
  );

  function renderInfo() {
    if (info === 'none') {
      return;
    }

    const ProgressIcon = {
      success: <CheckCircleOutlineIcon height={16} width={16} />,
      error: <ErrorOutlineIcon height={16} width={16} />,
    } as const;

    let content = (
      <ProgressBarPercentageTextStyled state={state}>
        {`${normalizedValue}%`}
      </ProgressBarPercentageTextStyled>
    );

    if (state !== 'active') {
      if (info === 'icon') {
        content = (
          <ProgressBarIconStyled state={state}>
            {ProgressIcon[state]}
          </ProgressBarIconStyled>
        );
      }

      if (info === 'both') {
        content = (
          <>
            {content}

            <ProgressBarIconStyled state={state}>
              {ProgressIcon[state]}
            </ProgressBarIconStyled>
          </>
        );
      }
    }

    return content;
  }
};
