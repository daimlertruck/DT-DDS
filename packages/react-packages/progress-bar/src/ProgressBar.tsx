import { Icon } from '@dt-ui/react-icon';
import { Typography } from '@dt-ui/react-typography';
import { useTheme } from '@emotion/react';

import { MAX_PROGRESS_BAR_VALUE, MIN_PROGRESS_BAR_VALUE } from './constants';
import {
  ProgressBarFieldStyled,
  ProgressBarWrapperStyled,
  ProgressBarStyled,
  ProgressBarTrackStyled,
  ProgressBarIconStyled,
  ProgressBarPercentageTextStyled,
  ProgressBarHelperTextStyled,
} from './ProgressBar.styled';
import { ProgressBarSize, ProgressBarState, ProgressBarInfo } from './types';

export interface ProgressBarProps {
  state: ProgressBarState;
  value: number;
  label?: string;
  helperText?: string;
  info?: ProgressBarInfo;
  size?: ProgressBarSize;
}

export const ProgressBar = ({
  state,
  value,
  label,
  helperText,
  size = 'small',
  info = 'none',
}: ProgressBarProps) => {
  const normalizedValue = Math.min(
    Math.max(value, MIN_PROGRESS_BAR_VALUE),
    MAX_PROGRESS_BAR_VALUE
  );
  const theme = useTheme();

  return (
    <ProgressBarFieldStyled>
      {label ? (
        <Typography color='content.body' fontStyles='body2Bold'>
          {label}
        </Typography>
      ) : null}

      <ProgressBarWrapperStyled>
        <ProgressBarTrackStyled
          aria-label={label}
          aria-valuemax={MAX_PROGRESS_BAR_VALUE}
          aria-valuemin={MIN_PROGRESS_BAR_VALUE}
          aria-valuenow={normalizedValue}
          role='progressbar'
        >
          <ProgressBarStyled
            progress={normalizedValue}
            size={size}
            state={state}
          />
        </ProgressBarTrackStyled>

        {renderInfo()}
      </ProgressBarWrapperStyled>

      {helperText ? (
        <ProgressBarHelperTextStyled state={state}>
          {helperText}
        </ProgressBarHelperTextStyled>
      ) : null}
    </ProgressBarFieldStyled>
  );

  function renderInfo() {
    if (info === 'none') {
      return;
    }

    const ProgressIcon = {
      success: (
        <Icon
          code='check_circle'
          color={theme.palette.success.default}
          size='medium'
        />
      ),
      error: (
        <Icon code='error' color={theme.palette.error.default} size='medium' />
      ),
    } as const;

    let content = (
      <ProgressBarPercentageTextStyled state={state}>
        {`${normalizedValue}%`}
      </ProgressBarPercentageTextStyled>
    );

    if (state !== 'active') {
      if (info === 'icon') {
        content = (
          <ProgressBarIconStyled>{ProgressIcon[state]}</ProgressBarIconStyled>
        );
      }

      if (info === 'both') {
        content = (
          <>
            {content}
            <ProgressBarIconStyled>{ProgressIcon[state]}</ProgressBarIconStyled>
          </>
        );
      }
    }

    return content;
  }
};
