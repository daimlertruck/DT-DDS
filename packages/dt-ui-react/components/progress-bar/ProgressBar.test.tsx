import { withProviders } from '@dt-ui/react-core';
import { render } from '@testing-library/react';

import { ProgressBar } from './ProgressBar';

describe('<ProgressBar /> component', () => {
  const ProvidedProgressBar = withProviders(ProgressBar);

  it('should render the default options correctly', () => {
    const { container } = render(
      <ProvidedProgressBar state='active' value={40} />
    );

    expect(container).toMatchSnapshot();
  });

  describe('should render the sizes correctly', () => {
    it('should render the small size correctly when the prop is provided', () => {
      const { container } = render(
        <ProvidedProgressBar size='small' state='active' value={40} />
      );

      expect(container).toMatchSnapshot();
    });

    it('should render the large size correctly when the prop is provided', () => {
      const { container } = render(
        <ProvidedProgressBar size='large' state='active' value={40} />
      );

      expect(container).toMatchSnapshot();
    });
  });

  describe('should render the states correctly', () => {
    it('should render the success state correctly when the prop is provided', () => {
      const { container } = render(
        <ProvidedProgressBar state='success' value={40} />
      );

      expect(container).toMatchSnapshot();
    });

    it('should render the error state correctly when the prop is provided', () => {
      const { container } = render(
        <ProvidedProgressBar state='error' value={40} />
      );

      expect(container).toMatchSnapshot();
    });
  });

  it('should render the percentage', () => {
    const { container } = render(
      <ProvidedProgressBar state='active' value={40} />
    );

    expect(container).toMatchSnapshot();
  });

  it('should render the label when prop is provided', () => {
    const { container } = render(
      <ProvidedProgressBar
        label='Progress Bar Label'
        state='active'
        value={40}
      />
    );

    expect(container).toMatchSnapshot();
  });

  it('should render the helper text when prop is provided', () => {
    const { container } = render(
      <ProvidedProgressBar
        helperText='Progress Bar Helper'
        state='active'
        value={40}
      />
    );

    expect(container).toMatchSnapshot();
  });
});
