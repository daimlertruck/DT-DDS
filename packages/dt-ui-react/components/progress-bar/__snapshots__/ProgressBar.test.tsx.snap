// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ProgressBar /> component should render the default options correctly 1`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: flex-start;
  -webkit-box-align: flex-start;
  -ms-flex-align: flex-start;
  align-items: flex-start;
  gap: 4px;
  min-width: 48px;
}

.emotion-2 {
  display: grid;
  grid-template-columns: minmax(48px, 1fr) 28px 16px;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  gap: 4px;
  width: 100%;
  height: 16px;
}

.emotion-4 {
  width: 100%;
  background-color: #EBECED;
  border-radius: 0px;
  overflow: hidden;
}

.emotion-6 {
  width: 40%;
  height: 4px;
  background-color: #008DFC;
  -webkit-transition: width 0.2s ease-in-out;
  transition: width 0.2s ease-in-out;
  border-radius: 0px 0px 0px 0px;
}

<div>
  <div
    class="emotion-0 emotion-1"
  >
    <div
      class="emotion-2 emotion-3"
    >
      <div
        aria-valuemax="100"
        aria-valuemin="0"
        aria-valuenow="40"
        class="emotion-4 emotion-5"
        role="progressbar"
      >
        <div
          class="emotion-6 emotion-7"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`<ProgressBar /> component should render the helper text when prop is provided 1`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: flex-start;
  -webkit-box-align: flex-start;
  -ms-flex-align: flex-start;
  align-items: flex-start;
  gap: 4px;
  min-width: 48px;
}

.emotion-2 {
  display: grid;
  grid-template-columns: minmax(48px, 1fr) 28px 16px;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  gap: 4px;
  width: 100%;
  height: 16px;
}

.emotion-4 {
  width: 100%;
  background-color: #EBECED;
  border-radius: 0px;
  overflow: hidden;
}

.emotion-6 {
  width: 40%;
  height: 4px;
  background-color: #008DFC;
  -webkit-transition: width 0.2s ease-in-out;
  transition: width 0.2s ease-in-out;
  border-radius: 0px 0px 0px 0px;
}

.emotion-8 {
  font-weight: 400;
  font-size: 0.75rem;
  line-height: 1rem;
  letter-spacing: normal;
  margin: 0;
  text-transform: none;
  color: #A1A2A3;
}

<div>
  <div
    class="emotion-0 emotion-1"
  >
    <div
      class="emotion-2 emotion-3"
    >
      <div
        aria-valuemax="100"
        aria-valuemin="0"
        aria-valuenow="40"
        class="emotion-4 emotion-5"
        role="progressbar"
      >
        <div
          class="emotion-6 emotion-7"
        />
      </div>
    </div>
    <p
      class="emotion-8 emotion-9"
    >
      Progress Bar Helper
    </p>
  </div>
</div>
`;

exports[`<ProgressBar /> component should render the label when prop is provided 1`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: flex-start;
  -webkit-box-align: flex-start;
  -ms-flex-align: flex-start;
  align-items: flex-start;
  gap: 4px;
  min-width: 48px;
}

.emotion-2 {
  font-weight: 700;
  font-size: 0.875rem;
  line-height: 1.25rem;
  letter-spacing: normal;
  margin: 0;
  text-transform: none;
  color: #636464;
}

.emotion-4 {
  display: grid;
  grid-template-columns: minmax(48px, 1fr) 28px 16px;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  gap: 4px;
  width: 100%;
  height: 16px;
}

.emotion-6 {
  width: 100%;
  background-color: #EBECED;
  border-radius: 0px;
  overflow: hidden;
}

.emotion-8 {
  width: 40%;
  height: 4px;
  background-color: #008DFC;
  -webkit-transition: width 0.2s ease-in-out;
  transition: width 0.2s ease-in-out;
  border-radius: 0px 0px 0px 0px;
}

<div>
  <div
    class="emotion-0 emotion-1"
  >
    <p
      class="emotion-2 emotion-3"
      color="content.body"
      data-testid="typography"
    >
      Progress Bar Label
    </p>
    <div
      class="emotion-4 emotion-5"
    >
      <div
        aria-label="Progress Bar Label"
        aria-valuemax="100"
        aria-valuemin="0"
        aria-valuenow="40"
        class="emotion-6 emotion-7"
        role="progressbar"
      >
        <div
          class="emotion-8 emotion-9"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`<ProgressBar /> component should render the percentage 1`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: flex-start;
  -webkit-box-align: flex-start;
  -ms-flex-align: flex-start;
  align-items: flex-start;
  gap: 4px;
  min-width: 48px;
}

.emotion-2 {
  display: grid;
  grid-template-columns: minmax(48px, 1fr) 28px 16px;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  gap: 4px;
  width: 100%;
  height: 16px;
}

.emotion-4 {
  width: 100%;
  background-color: #EBECED;
  border-radius: 0px;
  overflow: hidden;
}

.emotion-6 {
  width: 40%;
  height: 4px;
  background-color: #008DFC;
  -webkit-transition: width 0.2s ease-in-out;
  transition: width 0.2s ease-in-out;
  border-radius: 0px 0px 0px 0px;
}

<div>
  <div
    class="emotion-0 emotion-1"
  >
    <div
      class="emotion-2 emotion-3"
    >
      <div
        aria-valuemax="100"
        aria-valuemin="0"
        aria-valuenow="40"
        class="emotion-4 emotion-5"
        role="progressbar"
      >
        <div
          class="emotion-6 emotion-7"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`<ProgressBar /> component should render the sizes correctly should render the large size correctly when the prop is provided 1`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: flex-start;
  -webkit-box-align: flex-start;
  -ms-flex-align: flex-start;
  align-items: flex-start;
  gap: 4px;
  min-width: 48px;
}

.emotion-2 {
  display: grid;
  grid-template-columns: minmax(48px, 1fr) 28px 16px;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  gap: 4px;
  width: 100%;
  height: 16px;
}

.emotion-4 {
  width: 100%;
  background-color: #EBECED;
  border-radius: 0px;
  overflow: hidden;
}

.emotion-6 {
  width: 40%;
  height: 8px;
  background-color: #008DFC;
  -webkit-transition: width 0.2s ease-in-out;
  transition: width 0.2s ease-in-out;
  border-radius: 0px 0px 0px 0px;
}

<div>
  <div
    class="emotion-0 emotion-1"
  >
    <div
      class="emotion-2 emotion-3"
    >
      <div
        aria-valuemax="100"
        aria-valuemin="0"
        aria-valuenow="40"
        class="emotion-4 emotion-5"
        role="progressbar"
      >
        <div
          class="emotion-6 emotion-7"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`<ProgressBar /> component should render the sizes correctly should render the small size correctly when the prop is provided 1`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: flex-start;
  -webkit-box-align: flex-start;
  -ms-flex-align: flex-start;
  align-items: flex-start;
  gap: 4px;
  min-width: 48px;
}

.emotion-2 {
  display: grid;
  grid-template-columns: minmax(48px, 1fr) 28px 16px;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  gap: 4px;
  width: 100%;
  height: 16px;
}

.emotion-4 {
  width: 100%;
  background-color: #EBECED;
  border-radius: 0px;
  overflow: hidden;
}

.emotion-6 {
  width: 40%;
  height: 4px;
  background-color: #008DFC;
  -webkit-transition: width 0.2s ease-in-out;
  transition: width 0.2s ease-in-out;
  border-radius: 0px 0px 0px 0px;
}

<div>
  <div
    class="emotion-0 emotion-1"
  >
    <div
      class="emotion-2 emotion-3"
    >
      <div
        aria-valuemax="100"
        aria-valuemin="0"
        aria-valuenow="40"
        class="emotion-4 emotion-5"
        role="progressbar"
      >
        <div
          class="emotion-6 emotion-7"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`<ProgressBar /> component should render the states correctly should render the error state correctly when the prop is provided 1`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: flex-start;
  -webkit-box-align: flex-start;
  -ms-flex-align: flex-start;
  align-items: flex-start;
  gap: 4px;
  min-width: 48px;
}

.emotion-2 {
  display: grid;
  grid-template-columns: minmax(48px, 1fr) 28px 16px;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  gap: 4px;
  width: 100%;
  height: 16px;
}

.emotion-4 {
  width: 100%;
  background-color: #EBECED;
  border-radius: 0px;
  overflow: hidden;
}

.emotion-6 {
  width: 40%;
  height: 4px;
  background-color: #FF4A4A;
  -webkit-transition: width 0.2s ease-in-out;
  transition: width 0.2s ease-in-out;
  border-radius: 0px 0px 0px 0px;
}

<div>
  <div
    class="emotion-0 emotion-1"
  >
    <div
      class="emotion-2 emotion-3"
    >
      <div
        aria-valuemax="100"
        aria-valuemin="0"
        aria-valuenow="40"
        class="emotion-4 emotion-5"
        role="progressbar"
      >
        <div
          class="emotion-6 emotion-7"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`<ProgressBar /> component should render the states correctly should render the success state correctly when the prop is provided 1`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: flex-start;
  -webkit-box-align: flex-start;
  -ms-flex-align: flex-start;
  align-items: flex-start;
  gap: 4px;
  min-width: 48px;
}

.emotion-2 {
  display: grid;
  grid-template-columns: minmax(48px, 1fr) 28px 16px;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  gap: 4px;
  width: 100%;
  height: 16px;
}

.emotion-4 {
  width: 100%;
  background-color: #EBECED;
  border-radius: 0px;
  overflow: hidden;
}

.emotion-6 {
  width: 40%;
  height: 4px;
  background-color: #21A330;
  -webkit-transition: width 0.2s ease-in-out;
  transition: width 0.2s ease-in-out;
  border-radius: 0px 0px 0px 0px;
}

<div>
  <div
    class="emotion-0 emotion-1"
  >
    <div
      class="emotion-2 emotion-3"
    >
      <div
        aria-valuemax="100"
        aria-valuemin="0"
        aria-valuenow="40"
        class="emotion-4 emotion-5"
        role="progressbar"
      >
        <div
          class="emotion-6 emotion-7"
        />
      </div>
    </div>
  </div>
</div>
`;
